docs/RBAC.md

```diff
@@ -0,0 +1,97 @@
+# Role Based Access Control (RBAC)
+
+### Status: DRAFT
+**This feature is not fully implemented. This document describes the desired imlpementation for Odessey (search v2).**
+
+The search service collects data using a service account with wide cluster access and stores all resources in the database. The API must enforce that results for each user (or service account) only contain resources that they are authorized to access.
+
+## Access to the API
+<!-- This feature is new for V2 -->
+The API itself is protected by RBAC. Users must be given a role that allows access to search.
+
+The default ACM admin and viewer roles should include access to the search API by default. [TODO: Describe the roles and what is added.]
+
+> **DISCUSSION:** Implementation options.
+> 1. Management Ingress? No, this will be deprecated.
+> 2. Kube API server extension?
+> 3. Admission webhook?
+> 4. Validation at the service? I want to avoid this if possible.
+
+## Enforcing RBAC on results
+
+The API authenticates the user (or service account) and impersonates the user to obtain their access rules.
+
+> **DISCUSSION:** What's the correct way to impersonate the user? Currently, we are just using their token, which isn't ideal. The request should made "by search on behalf of user".
+> 
+> I expect this to be similar to authorizing an app on Github to take actions on my behalf.
+
```

Comment:
There are 2 scenarios in the long run. They may not be applicable today. See if you concur:
1. user makes a synchronous call and needs an answer. The user token is available
2. Another service is making calls in the background may be on behalf of a user request that was initiated a while back. There is no token in this case (we may not have a case for this now - but it usually one of these comes up specially we come up with long running API - given the fact that we have to gather data from a large DB). We will have to trust the service in this case.
In other words, no matter what the implementation is, we will have to evaluate the rights of the user and perhaps also rights of other services who may call us w/o any user credentials. Does this make sense?

Reply:
Ok, let's talk scenario #2.  I believe that we can cover it using impersonation.  We'll need to authenticate and trust both the App and the user, then impersonate the user.

https://kubernetes.io/docs/reference/access-authn-authz/authentication/#user-impersonation